### PR TITLE
more useful test output reporting

### PR DESCRIPTION
**Pull request by tarballs-are-good**
This PR significantly changes what the output looks like for running tests. It is noisier, but a lot more useful during actual Lisp development, or even when reading logs from a CI server.

At Rigetti Computing, it has been very frequently annoying to have to look at the backtrace to see what test has failed during an interactive session. It is also extremely confusing to notice which test is printing intermediate output.

This change changes from the old look:

```
PROJECT-TESTS (Suite)
  TEST-QUEUE                                                             [ OK ]
  TEST-QUEUE-MAKE                                                        [ OK ]
  TEST-NSPLIT                                                            [ OK ]
  TEST-LEXER                                                             [ OK ]
    Testing good file bell
    Testing good file empty
  TEST-PARSING-GOOD-TEST-FILES                                           [ OK ]
```
Note that the intermediate output comes from `TEST-PARSING-GOOD...`!

The new, unconfusing look is:

```
CL-QUIL-TESTS (Suite)
  Starting TEST-QUEUE
  Finished TEST-QUEUE  . . . . . . . . . . . . . . . . . . . . . . . . . [ OK ]
  Starting TEST-QUEUE-MAKE
  Finished TEST-QUEUE-MAKE . . . . . . . . . . . . . . . . . . . . . . . [ OK ]
  Starting TEST-NSPLIT
  Finished TEST-NSPLIT . . . . . . . . . . . . . . . . . . . . . . . . . [ OK ]
  Starting TEST-LEXER
  Finished TEST-LEXER  . . . . . . . . . . . . . . . . . . . . . . . . . [ OK ]
  Starting TEST-PARSING-GOOD-TEST-FILES
    Testing good file bell
    Testing good file empty
  Finished TEST-PARSING-GOOD-TEST-FILES  . . . . . . . . . . . . . . . . [ OK ]
```

### CODE REVIEW

**Comment by luismbo** [comment]
I think we could do something like this with some stream-fu (basically a stream that keeps track of the current column). Do you think it'd be worth the trouble?

```
CL-QUIL-TESTS (Suite)
  TEST-LEXER  . . . . . . . . . . . . . . . . . . . . . . . . . . . . . [ OK ]
  TEST-PARSING-GOOD-TEST-FILES
    Testing good file a
    Testing good file b . . . . . . . . . . . . . . . . . . . . . . . . [ OK ]
```

**Comment by luismbo** [comment]
Another potential use of a "test log" stream could be the automatic timestamping of each line.

**Comment by joaotavora** [comment]
@luismbo I was going to ask about that, I think it's the better alternative.  I'm also not superfond of the dots, but I could learn to like them, or maybe a tiny option could be found for that.

**Comment by tarballs-are-good** [comment]
The dots are really helpful for very long test output. It's hard to distinguish which lines are OK/not OK in the right column. Maybe I just don't have great vision, but I have to track across with my finger sometimes.

**Comment by tarballs-are-good** [comment]
@joaotavora @luismbo What's the next step? Do we really want to be doing things like counting columns? Is that information even guaranteed with something like gray streams? It seems like overkill. Maybe we should just not print the test name when it says "FINISHED"?

**Comment by joaotavora** [comment]
> The dots are really helpful for very long test output.

OK, then maybe we can add them as an optional feature and then think about turning them on by default.

> Do we really want to be doing things like counting columns?

If it's easy to do, I don't see why not.  It brings, as Luis pointed, added benefits like timestamping possibilities.  

The dependency on `trivial-gray-streams` seems, well, trivial. 

> Is that information even guaranteed with something like gray streams?

My understanding is that we can hook onto writing functions and do the counting outselves.

> It seems like overkill.

The same could be said of your proposal to solve what could be seen as a minor nit. After all I don't think there is any ambiguity in the output. Emacs's `ert` framework, powerful as it is, also suffers from the same problem (and I don't know any other that doesn't).

To summarize: 

1. I think @luismbo 's approach is better.  I'm wrapping my head around `trivial-gray-streams`.  It would seem all that is needed is a custom class with a column counter and an "understream". This custom class would be used in `FIASCO:RUN-SUITE-TESTS` to set `FIASCO:*PRETTY-LOG-STREAM*`.
2. If we can't do 1., we could implement yours and make it optional via a switch.

What do you think?

**Comment by joaotavora** [comment]
Fiasco's self tests for the commit  running in Travis are a good indication of how it works. Other than that, it has very light testing.  I didn't put the dots in yet (that is trivial I think).

https://travis-ci.org/joaotavora/fiasco/builds/425020182?utm_source=github_status&utm_medium=notification

**Comment by tarballs-are-good** [comment]
@joaotavora Sorry I'm late on this. I'll take a look.

**Comment by joaotavora** [comment]
> The dots are slightly non-trivial if you want them to look nice and have them aligned

Ok. Feel free to put that part in as another commit on top of mine

**Comment by joaotavora** [comment]
> Does that affect the new stream type?

No idea, left as an exercise for the reader. 

But maybe not. LF, which I suppose is the one that translates to `#\Newline`, comes after, comes after CR

**Comment by tarballs-are-good** [comment]
@joaotavora So to recap: Your current ask is that I re-implement the contents of this PR atop your branch?

**Comment by joaotavora** [comment]
> @joaotavora So to recap: Your current ask is that I re-implement the contents of this PR atop your branch?

Err, in a way, yes.  But let's recap yes: I proposed the column-counting strategy as a way to:

1. Solve the problem you descried
2. Do it in a way that doesn't change parts I like about fiasco's test reporting.

If you agree with my proposal, we should:

a. Address your code-style comments
b. Add the dots for readability, unless @luismbo opposes

One good way to do this is to make `a` a commit and `b` another commit, both on top of https://github.com/joaotavora/fiasco/commit/160dd3fad507532a7a31a4a7ac31a397f9901146. Then rebase this PR to that commit for a final review and merging.

If you can do both `a` and `b`, that's great, otherwise give me some time and I'll do it (though I can't touch the branch connected to the PR: you'll have to that that bit)

**Comment by joaotavora** [comment]
Hi @tarballs-are-good 

I pushed the changes myself, closing this PR and hopefully addressing your comments.  Let me know if you like it.

Thanks
